### ISSUE TITLE
publish: deleting group associated with notebook removes said notebook from your interface, but does not remove it from your state

### ISSUE THREAD

**Issue by ixv** [urbit/urbit#2856]
If you have a notebook associated with a group, and the group gets deleted, the notebook will continue to exist in state. However on the frontend, for managed notebooks we require the presence of an actual group in order to render it.

This seems like one of those things that will need to be addressed by our "archived content" story. Not sure if there's a suitable near-term solution for this without that. Note that this has some funky downstream effects, such as if you have unread notes in that notebook, they will be counted toward the tile total.

cc-ing @matildepark @loganallenc @urcades @Fang- I'm curious to get your takes on this.

**Comment by loganallenc** [comment]
I think we should mirror chat’s behavior in the meantime and delete the notebook.

**Comment by urcades** [comment]
Check my response via Tlon General:

<img width="1072" alt="Screen Shot 2020-05-05 at 5 53 52 PM" src="https://user-images.githubusercontent.com/1195363/81119803-86ee8180-8ef9-11ea-86c7-ec2784518a39.png">

tl;dr I basically agree w/ logan we should delete the channel in the near term (which will suck for some people, possibly), but we should plan out how to extract data post-deletion at some point.

**Comment by ixv** [comment]
Ah interesting, I didn't know chats were deleted in that case. I assumed it always took the policy of preserving local data.

**Comment by Fang-** [comment]
> @Fang- I'm curious to get your takes on this.

I feel pretty negatively about the eager deletion behavior discussed above. It's data that lives on _my_ ship, so it's _my_ data, and that shouldn't ever be deleted (at least in large amounts like this) without my explicit consent, or at least giving me the chance to back it up.

**Comment by matildepark** [comment]
@Fang- see [here](https://github.com/urbit/urbit/pull/2740#issuecomment-614164215). Planning around archiving is its own slate of work that is itself not even started. The case should be with parity for Chat and if you feel strongly then advocate for higher priority for a holistically considered implementation. We shouldn't have half-implementations that differ per app.

**Comment by baudtack** [comment]
I'm with @Fang- You **do not** get to delete things from my ship. Ever.

**Comment by ixv** [comment]
Uh, @loganallenc I just tested this for chat and it does not work the way you said. The chat continues to exist and the chat tile-number has the same issue.

**Comment by eglaysher** [comment]
Deleting things from my ship is a no-no. Removing a subscription doesn't imply removing the data.

**Comment by liam-fitzgerald** [comment]
@ixv I suspect deleting the group never updates the permissions and so the backlog subscription is never kicked. Fwiw, chat deletes a room when you *leave* it's associated group, but it appears the deleting group case is bugged

**Comment by matildepark** [comment]
> I suspect deleting the group never updates the permissions and so the backlog subscription is never kicked.

We're going to investigate and file that as a separate issue, please stand by ...

**Comment by ixv** [comment]
fwiw, if chat also doesn't do this, I'm fine with a solution that just downgrades the chat/notebook to an unmanaged one.

**Comment by liam-fitzgerald** [comment]
@ixv deleting a groups deletes all the associated channels on my machine.  Are you running a version with #2841 in it?

**Comment by matildepark** [comment]
Are you talking about the case where a group is deleted from under you or when you delete your own group? There’s a few cases here…

— 
~haddef-sigwen
https://urbit.org

**Comment by liam-fitzgerald** [comment]
both

**Comment by ixv** [comment]
Just tested it again, on latest master with that PR in, still have the problem. To be clear, the issue is that the data still exists on the backend even though the chats/notebooks vanish from the frontend. Running `:chat-store +dbug %state` and `:publish +dbug %state` show they're still there.

**Comment by joemfb** [comment]
Adding my voice to @Fang-, @baudtack, @eglaysher, &c -- it's just not ok to delete data from my ship without explicit confirmation.

**Comment by ixv** [comment]
This really could be as simple as adding an "archived" flag in each chat/notebook and just having another sidebar section for them similar to "direct messages"/"unmanaged notebooks". Seems like a fine and easy to do interim solution until we work out the full archiving story.

**Comment by matildepark** [comment]
Nobody here is contesting the idealised state or the moral value of proper behaviour cross-ship.

> This seems like one of those things that will need to be addressed by our "archived content" story. Not sure if there's a suitable near-term solution for this without that.

Features need a story. And until then, we have to implement them the best we can, and we accept the tradeoff for now. Archiving content, displaying archived content, actually thinking about how to present archived content, whether it's exported, et al. are all essentially in design stages and were part of this sprint of work, but were deprioritised while we refactor all our userspace apps. We are a team of 4. Half of them can do back-end.

> This really could be as simple as adding an "archived" flag in each chat/notebook and just having another sidebar section for them similar to "direct messages"/"unmanaged notebooks". Seems like a fine and easy to do interim solution until we work out the full archiving story.

This then means that we have to spend some time doing front-end work and adding a feature while we refactor the front-end into an SPA, something that has been frozen. SPA is coming along fast, but again, we are not opening the gate to any and every feature to slip in. Time slips by; this bug persists. Why? Because we didn't do the simplest thing first.

Once we do this, what will we say? Archived notebooks exist after a group is gone, but chats delete themselves? OK, so now we do front-end and revise chat's behaviour too, and we test that. So now we have double the work to test. And what about Links? Etc. etc. etc.

Part of my job is to choreograph our feature slate to have a harmony and a rhythm across all our projects without blocking each other, and accept the tradeoffs.

Therefore, @loganallenc is right.

> I think we should mirror chat’s behavior in the meantime and delete the notebook.

This being "wrong" just means we prioritise it higher to actually do properly ... again, once we're done refactoring. This immoral behaviour already exists. Is already on your ship. And is an edge case with a limited scope.

**Comment by ixv** [comment]
Chat behavior is currently the same as publish!

**Comment by joemfb** [comment]
Framing this as an important part of our longer-term vision is *exactly* what's being contested. If there is code on my ship now that unconditionally deletes my data, that is unacceptable and *must* be dealt with immediately. This is not a feature request, but concerns the basic value proposition of all personal urbit use-cases.

Of course the overall archival story needs to be developed, and access to these things needs to fit neatly into the overall design of our interfaces. And that can't simply be demanded -- we're all dealing with limited resources. But there is a tremendous difference between important data on my ship being inconsistently available (or even unavailable) through the UI, and that data being deleted without recourse. The former is unfortunate, the latter is unacceptable.

**Comment by matildepark** [comment]
> But there is a tremendous difference between important data on my ship being inconsistently available (or even unavailable) through the UI, and that data being deleted without recourse. The former is unfortunate, the latter is unacceptable.

I find our UI showing incorrect information and creating a confounding user experience due to half-considered implementations from an abstract, infrastructure-first ideological position to be unacceptable. We are exclusively talking about resources that we obtained by subscribing to another ship. Who owns them? Technically, and ideally, we all do.

So why do we delete content? Because our groups model itself uses a "group owner" who decides who and what is in the group. It is, in its implementation, essentially client-server permissioned. And everything sprung from that; it is a wholly considered UX and back-end implementation pattern that has been debated to death since it was written months ago.

And it's *also* on the slate to be rewritten, and it's *also* coming after the refactor.

I'm really not in the mood to debate this. We have a roadmap that makes use of the project to its fullest. Refactor will come first. We are not leaving this bug hanging for an indefinite because of this.

**Comment by matildepark** [comment]
Talked around trying to get a sense of the cost and the detriments. 

Essentially AA feels we should postpone the decision until we can do the right thing. And because of the cost of adding and surfacing an archived param across FE being indeterminate, we’re going to leave Publish and Chat with wonky unread counts for now since their behaviour both don’t delete resources.

I’m gonna talk to @urcades about archiving story and slate it higher and continue as we were.

**Comment by urcades** [comment]
I want to retroactively amend my POV by stating that when I said "deletion" in my first response to this issue, I was mainly speaking to the idea of removing the interface-object of the channel from the interface, not "absolute deletion" in the form of removing data that exists on your ship. 

This might map to simply unsubscribing from a channel that was nested under a "just-deleted" group and visually removing it from one's list of channels.

**Comment by vvisigoth** [comment]
I've had a few conversations with people and I thought I'd put my thoughts here for clarity.

From my perspective there are two ideals to which we are committed:

1. You are sovereign over your ship (and its data)
2. The interface is a faithful representation of your ship's state

These two ideals are in tension right now, but a goal of this entire project/product is to harmonize them. Harmonizing these will allow us to offer groups not as a service, but as a subscription product. Renting v. a mortgage. This should excite us all. But our dispute is _how_ to harmonize them. 

My position is that we are much closer to 1 than 2 (we've been at it for longer) and that jeopardizing 1 has a higher cost than jeopardizing 2. Jeopardizing 1 looks like deleting something because you want to de-emphasize it in the UI and jeopardizing 2 looks like having a slightly janky FE codebase because you're introducing more heterogeneity. The latter option has several smaller decision points that should be evaluated on a cost-benefit basis. I have confidence that the OS1 team can figure that out.

<@vvisigoth sings we shall overcome>

**Comment by matildepark** [comment]
You have no idea how much I don't want to keep talking about this.

Essentially where OS1 team is at is that "adding an archived param and janky FE" as a distinct unit of work, consistently across our applications, has an as-yet-indeterminate cost to implement (it's severely complicated by how Publish and Chat differently manifest unread counts in tiles/FE); we are focusing on other things right now, refactoring bespoke userspace applications into predictable, clean patterns, and given that Chat *doesn't* delete resources due to some other bug, both "delete stuff" and "hide stuff" have costs. And both probably cost enough that it would be basically the same cost as just doing it properly!

I talked to @loganallenc; I talked to @ixv; I talked to @vvisigoth; essentially it puts us in a tight bind where the broken unread counts are just blocked on this issue, and we're going to tackle it properly after SPA lands and when the back-end work makes sense to start thinking about it. I've slated "archived content" for design consideration stage and we'll just push it up a little, as previously stated.

**Comment by ohAitch** [comment]
Ultimately the question is, if a chat is deleted now, will we be able to retrieve it in three years: if it's inaccessible from the current UI but there, the answer will by then be yes, while if it's deleted on the backend(modulo event log spelunking), the data is lost.

Which imo suggests a hack on top of a hack: whenever a room is unsubscribed / removed, snapshot the entire chat state(or ideally some subset), then delete it from the "primary" store key-value everything else is using for unread counts &c. Copy-on-write datastructures should make this non-prohibitive, and let you always pull it out later once an archive system is in place(or by adding a "dump all chat messages we have ever seen to clay" admin poke), while leaving open the critical path of "don't try to shove a schema migration into the project schedule"

**Comment by loganallenc** [comment]
We don’t need a schema migration to institute this, and it will be a relatively light change. Thanks for your comment.

**Comment by matildepark** [comment]
Ancient thread. The problem of archived content has been punted way until, well, now. This is of course really not acceptable because it led to UX being inexplicably "broken" for ages. 

On behalf of the team, I apologise for all those who had to deal with this invisible monster. It had hundreds of heads: any time any user who hosted, or was the only other participant in a resource breached; kicked; deleted, or what have you, 

- Things either inexplicably "didn't work", letting you submit content with a silent fail; or 
- Worse, inexplicably "did work", showing that you submitted successfully, even though *there was nobody else there*, and they had their own copy of the content, locked in the Shadow Realm.

The current state of this issue as it stands is as follows:

- For Links and Publish, as they are graph resources, which was written with archives in mind, this is solved as of #3986. When you leave or are kicked, we clean up your state and keep an archived version. The archived graphs live in your ship's state; they can be exported. We don't serve this content to the front-end, and likewise there is no front-end affordance to look at archived content ... but that's a feature request we can get to.
- For Chat, as the lone non-graph store still remaining, the conversation lives in the symptomatic #3926. This is far more critical, but we'll prioritise it in the short-term to settle.

As written in the OP, this issue in Publish is fixed as of #3473.
